turbopuffer/namespace.py

```diff
@@ -31,7 +31,7 @@ def __str__(self) -> str:
         return f'tpuf-namespace:{self.name}'
 
     @overload
-    def upsert(self, ids: List[int], vectors: List[List[float]], attributes: Optional[Dict[str, List[Optional[str]]]] = None) -> None:
+    def upsert(self, ids: Union[List[int], List[str]], vectors: List[List[float]], attributes: Optional[Dict[str, List[Optional[str]]]] = None) -> None:
```

Comment:
I like how the column oriented API makes it clear the IDs need to all be the same type

Reply:
Yep, that was a conscious decision to use over `List[Union[int, str]]`. The code doesn't enforce these annotations, but some client linters might.